File: ruby/parser.rb
# Stream chars one by one out of an input string
class InputStream
  attr_reader :input

  def initialize(input)
    @input = input
    @pos = 0
    @line = 1
    @col = 0
  end

  def next
    ch = @input[@pos]
    @pos += 1
    if ch == '\n'
      @line += 1
      @col = 0
    else
      @col += 1
    end
    ch
  end

  def peek
    return unless @pos < @input.length
    @input[@pos]
  end

  def eof
    peek.nil?
  end

  def croak(msg)
    raise Error("#{msg}, #{@line}:#{@col}")
  end
end

# Stream tokens one by one out of an input stream
class TokenStream
  def initialize(input)
    @input = input
  end

  def croak
    @input.croak
  end

  def next
    tok = @current
    @current = nil
    tok || read_next
  end

  private

  def read_while
    string = ''
    string += @input.next while !@input.eof && yield(@input.peek)
    string
  end

  def read_escaped
    escaped = false
    string = ''
    @input.next
    until @input.eof
      ch = @input.next
      if escaped
        string += ch
        escaped = false
      elsif ch == '\\'
        escaped = true
      elsif ch == '"'
        break
      else
        string += ch
      end
    end
    string
  end

  def read_string
    { type: :symbol, value: read_escaped }
  end

  def read_symbol
    identifier = read_while { |c| !" \t\n()".include?(c) }
    { type: :symbol, value: identifier }
  end

  def skip_comment
    read_while { |c| c != '\n' }
    @input.next
  end

  def read_next
    read_while { |c| " \t\n".include?(c) }
    return if @input.eof
    case @input.peek
    when ';'
      skip_comment
      read_next
    when '"'
      read_string
    when '(', ')'
      { type: :punc, value: @input.next }
    else
      read_symbol
    end
  end
end

# Parse code
module Parser
  class << self
    def parse(code)
      build_ast(TokenStream.new(InputStream.new(code)))
    end

    def print_ast(ast)
      print '('
      ast.each do |node|
        if node.is_a?(Array)
          print_ast(node)
        else
          print node
          print ' '
        end
      end
      print ') '
    end

    private

    def build_ast(tokstream)
      ast = []
      tok = tokstream.next
      while not_closed?(tok)
        if tok[:type] == :symbol
          ast.push(tok[:value])
        elsif tok[:type] == :punc && tok[:value] == '('
          ast.push(build_ast(tokstream))
        end
        tok = tokstream.next
      end
      ast
    end

    def not_closed?(tok)
      !tok.nil? && !(tok[:type] == :punc && tok[:value] == ')')
    end
  end
end

File: c++/parser.cc
#include "parser.hh"

Ast::Ast(std::string symbol) {
	type = VAR;
	var = symbol;
}

Ast::Ast(std::vector<Ast>* list_) {
	type = LIST;
	list = list_;
}

class InputStream
{
	std::stringstream* code;
public:
	InputStream(std::stringstream* code_)
	{
		code = code_;
	}

	char next()
	{
		return code->get();
	}

	char peek()
	{
		return code->peek();
	}

	bool eof()
	{
		return peek() == EOF;
	}
};

class Token
{
public:
	enum TokenType {
		Open,
		Close,
		Symbol
	} type;
	std::string symbol;

	Token(TokenType type_)
	{
		type = type_;
	}
	Token(std::string symbol_)
	{
		type = Symbol;
		symbol = symbol_;
	}
};

bool isWhitespace(char ch)
{
	return ch == ' ' || ch == '\t' || ch == '\n';
}

bool isParenthesis(char ch) {
	return ch == '(' || ch == ')';
}

bool isSymbol(char ch)
{
	return !isParenthesis(ch) && !isWhitespace(ch);
}

class TokenStream
{
	InputStream* input;

	std::string readWhile(bool (*predicate)(char))
	{
		std::string str;
		while (!input->eof() && predicate(input->peek())) {
			str += input->next();
		}
		return str;
	}

	void skipComment()
	{
		readWhile([](char c) {
			return c != '\n';
		});
		input->next();
	}

	std::string readEscaped()
	{
		bool escaped = false;
		std::string str;
		input->next();
		while (!input->eof()) {
			char c = input->next();
			if (escaped) {
				str += c;
				escaped = false;
			} else if (c == '\\') {
				escaped = true;
			} else if (c == '"') {
				break;
			} else {
				str += c;
			}
		}
		return str;
	}

	Token* readString()
	{
		std::string str = readEscaped();
		return new Token(str);
	}

	Token* readSymbol()
	{
		std::string str = readWhile(isSymbol);
		return new Token(str);
	}

	Token* readParenthesis()
	{
		char c = input->next();
		if (c == '(') {
			return new Token(Token::Open);
		} else {
			return new Token(Token::Close);
		}
	}

	Token* next()
	{
		readWhile(isWhitespace);
		if (input->eof()) {
			return NULL;
		}
		char c = input->peek();
		if (c == ';') {
			skipComment();
			return next();
		}
		if (c == '"') {
			return readString();
		}
		if (isSymbol(c)) {
			return readSymbol();
		}
		if (isParenthesis(c)) {
			return readParenthesis();
		}
		throw "I should not be here";
	}
public:
	TokenStream(std::stringstream* code)
	{
		input = new InputStream(code);
	}

	std::vector<Ast>* toAst()
	{
		std::vector<Ast>* ast = new std::vector<Ast>();
		Token* tok = next();
		while (tok != NULL && tok->type != Token::Close) {
			if (tok->type == Token::Symbol) {
				Ast* val = new Ast(tok->symbol);
				ast->push_back(*val);
			} else if (tok->type == Token::Open) {
				Ast* val = new Ast(toAst());
				ast->push_back(*val);
			}
			tok = next();
		}
		return ast;
	}
};


std::vector<Ast>* parseCode(std::stringstream& code)
{
	TokenStream tokenStream(&code);
	return tokenStream.toAst();
}


Ast::Ast(std::stringstream& code)
{
	type = LIST;
	list = parseCode(code);
}

Ast::~Ast()
{
	if (type == LIST) {
		delete list;
	}
}

std::ostream& operator<<(std::ostream& os, const Ast& ast)
{
	if (ast.type == Ast::VAR) {
		os << '"' << ast.var << "\" ";
	} else {
		os << " (";
		for (auto const& value: *ast.list) {
			os << value;
		}
		os << ") ";
	}
	return os;
}
